core/core-impl/src/main/java/io/openshift/appdev/missioncontrol/core/impl/MissionControlImpl.java

```diff
@@ -125,81 +146,14 @@ public Boom launch(final ForkProjectile projectile) throws IllegalArgumentExcept
-                log.log(Level.SEVERE, "Error while replacing README.adoc variables", e);
-            }
-        }
-
-        gitHubService.push(gitHubRepository, path);
-        statusEvent.fire(new StatusMessageEvent(projectile.getId(), StatusMessage.GITHUB_PUSHED));
-
-        OpenShiftProject openShiftProject = openShiftService.findProject(projectName).orElseGet(() -> openShiftService.createProject(projectName));
-        statusEvent.fire(new StatusMessageEvent(projectile.getId(), StatusMessage.OPENSHIFT_CREATE, singletonMap("location", openShiftProject.getConsoleOverviewUrl())));
-
-        List<AppInfo> apps = findProjectApps(path);
-        if (apps.isEmpty()) {
-            // Use Jenkins pipeline build
-            openShiftService.configureProject(openShiftProject, gitHubRepository.getGitCloneUri());
-        } else {
-            // Use S2I builder templates
-            for (AppInfo app : apps) {
-                for (File tpl : app.resources) {
-                    applyTemplate(openShiftService, gitHubRepository, openShiftProject, app, tpl);
-                }
-            }
-            for (AppInfo app : apps) {
-                for (File tpl : app.services) {
-                    applyTemplate(openShiftService, gitHubRepository, openShiftProject, app, tpl);
-                }
-            }
-            for (AppInfo app : apps) {
-                for (File tpl : app.apps) {
-                    applyTemplate(openShiftService, gitHubRepository, openShiftProject, app, tpl);
-                }
-            }
-        }
-        statusEvent.fire(new StatusMessageEvent(projectile.getId(), StatusMessage.OPENSHIFT_PIPELINE));
-
-        List<GitHubWebhook> webhooks = getGitHubWebhooks(gitHubService, openShiftService, gitHubRepository, openShiftProject);
-        statusEvent.fire(new StatusMessageEvent(projectile.getId(), StatusMessage.GITHUB_WEBHOOK));
-        launchEvent.fire(new LaunchEvent(getUserId(projectile), projectile.getId(), gitHubRepository.getFullName(), openShiftProject.getName(), projectile.getMission(), projectile.getRuntime()));
-        return new BoomImpl(gitHubRepository, openShiftProject, webhooks);
-    }
-
-    private void applyTemplate(OpenShiftService openShiftService, GitHubRepository gitHubRepository,
-            OpenShiftProject openShiftProject, AppInfo app, File tpl) {
-        try (FileInputStream fis = new FileInputStream(tpl)) {
-            openShiftService.configureProject(openShiftProject, fis, gitHubRepository.getGitCloneUri(), app.contextDir);
-        } catch (FileNotFoundException e) {
-            throw new IllegalStateException("Could not apply services template", e);
-        } catch (IOException e) {
-            throw new IllegalStateException("Could not read services template", e);
+        int startIndex = projectile.getStartOfStep();
+        for (int i = startIndex; i < commands.length; i++) {
+            Command command = commands[i];
+            command.execute(projectile);
```

Comment:
I wonder if we could use @Priority on top of @Observes methods to avoid introducing another framework

Reply:
what other framework are you referring to? I split the long method up into "commands" that execute a step that way it's easy to execute only a few.

Reply:
My idea was to avoid having "Commands" to be implemented and do that by observing events instead (eg. `void doSomething(@Observes @Priority(100) CreateProjectile event)`. 

It seems to be same use case as you did with the Status messages